### PR TITLE
[JENKINS-39363,CID-1205051] - Prevent resource leak in hudson.XmlFile#readRaw() in the case of encoding issues.

### PR DESCRIPTION
**Pull request by oleg-nenashev**
https://scan5.coverity.com/reports.htm#v36021/p10292/fileInstanceId=97573808&defectInstanceId=28155757&mergedDefectId=992726

@reviewbybees @jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/hudson/XmlFile.java [hunk 1]
```diff
@@ -51,6 +51,7 @@
 import java.io.Reader;
 import java.io.Writer;
 import java.io.StringWriter;
+import java.io.UnsupportedEncodingException;
```
Comment (olivergondza) [inline core/src/main/java/hudson/XmlFile.java]: Unused

File: core/src/main/java/hudson/XmlFile.java [hunk 2]
```diff
@@ -205,9 +206,20 @@ public String toString() {
...
     public Reader readRaw() throws IOException {
-        return new InputStreamReader(new FileInputStream(file),sniffEncoding());
+        FileInputStream fileInputStream = new FileInputStream(file);
+        try {
+            return new InputStreamReader(fileInputStream, sniffEncoding());
+        } catch(IOException ex) {
+            // Exception may happen if we fail to find encoding or if this encoding is unsupported.
+            // In such case we close the underlying stream and rethrow.
+            // If close() fails, we will throw its exception as the more important one.
+            fileInputStream.close();
```
Comment (olivergondza) [inline core/src/main/java/hudson/XmlFile.java]: > ... as the more important one.

Is there any particular reason why is it so? Usually, it is the less important one and attached as suppressed.

Comment (oleg-nenashev) [inline core/src/main/java/hudson/XmlFile.java]: It is kinda IMHO. But I can rework the code to give a priority to the original exception

**Comment by michaelneale** [comment]
:bee: (not sure if I agree with comment on which is the more important exception, perhaps original is most important, but not being able to close is also important).

File: core/src/main/java/hudson/XmlFile.java [hunk 3]
```diff
@@ -205,9 +206,20 @@ public String toString() {
      * Opens a {@link Reader} that loads XML.
      * This method uses {@link #sniffEncoding() the right encoding},
      * not just the system default encoding.
+     * @throws IOException Encoding issues
+     * @return Reader for the file. should be close externally once read.
      */
     public Reader readRaw() throws IOException {
-        return new InputStreamReader(new FileInputStream(file),sniffEncoding());
```
Comment (jtnord) [inline core/src/main/java/hudson/XmlFile.java]: seems to be the cause is using a string as a charset - just use a `Charset` so no streams are created unless you have a valid charset?

```java
String encoding = sniffEncoding();
  Charset charset = null;
try { 
  charset = Charset.forName(encoding);
}
catch (UnsupportedCharsetException | IllegalArgumentException | IllegalCharsetNameException ex) {
  throw new IOException("Failed to open file for reading", ex);
}
  FileInputStream fileInputStream = new FileInputStream(file);
  return new InputStreamReader(fileInputStream, charset);
```

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers Ready to go?  I'm totally confused by these PR builder results

**Comment by oleg-nenashev** [comment]
@kzantow Seems the PR builder has not been removed yet

File: core/src/main/java/hudson/XmlFile.java (hunk 3 above)
Comment (batmat) [inline core/src/main/java/hudson/XmlFile.java]: +1 and I'm not sure we should catch IO Exception inside, as it's already expected from the caller to close the stream.

**Comment by daniel-beck** [comment]
@oleg-nenashev He made it successful though, so it doesn't mark PRs unnecessarily failed. Not sure whether removing even works…

**Comment by daniel-beck** [comment]
@oleg-nenashev To clarify, this is done from your POV? There's a lengthy PR comment by @jtnord without response.

**Comment by oleg-nenashev** [comment]
Generally I agree with @jtnord, but it does not need to be a part of this fix since it's not 100% related. I'd prefer to follow-up in a separate PR

**Review by daniel-beck — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Merging according to the feedback.
Placeholder Jenkins issue: https://issues.jenkins-ci.org/browse/JENKINS-39363

**Comment by oleg-nenashev** [comment]
@jtnord Created https://issues.jenkins-ci.org/browse/JENKINS-39364 as a follow-up to your comment

**Pull request merged**